File: realtime/ios/webrtc/Sources/Auth/BackendJWTAuthProvider.swift
import Foundation

/// Fetches a short-lived JWT from a trusted backend (the JS example's `webrtc/jwt`
/// Node server) so no Inworld key/secret ships in the app.
struct BackendJWTAuthProvider: AuthProvider {
    var backendURL: String
    var session: URLSession = .shared

    private struct ConfigResponse: Decodable {
        var jwt: String
        var iceServers: [IceServer]?
        var url: String?

        enum CodingKeys: String, CodingKey {
            case jwt
            case iceServers = "ice_servers"
            case url
        }
    }

    func credentials() async throws -> RealtimeCredentials {
        guard let base = URL(string: backendURL) else { throw AuthError.invalidBackendURL }
        let configURL = base.appending(path: "api/config")
        let (data, response) = try await session.data(from: configURL)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw AuthError.backendRequestFailed("HTTP \(status)")
        }
        let config: ConfigResponse
        do {
            config = try JSONDecoder().decode(ConfigResponse.self, from: data)
        } catch {
            throw AuthError.backendRequestFailed("invalid config payload")
        }
        var callsURL: URL?
        if let urlString = config.url, !urlString.isEmpty {
            guard let parsed = URL(string: urlString) else {
                throw AuthError.backendRequestFailed("backend returned an invalid calls url: \(urlString)")
            }
            callsURL = parsed
        }
        return RealtimeCredentials(
            authorizationHeader: "Bearer \(config.jwt)",
            apiBaseURL: URL(string: "https://api.inworld.ai")!,
            callsURL: callsURL,
            preFetchedIceServers: config.iceServers
        )
    }
}

File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/auth/BackendJwtAuthProvider.kt
package ai.inworld.voiceagent.auth

import ai.inworld.voiceagent.realtime.Http
import ai.inworld.voiceagent.realtime.events.EventJson
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.net.URL

/** Fetches a short-lived JWT from a trusted backend (the JS example's `webrtc/jwt`
 *  Node server) so no Inworld key/secret ships in the app. */
class BackendJwtAuthProvider(private val backendUrl: String) : AuthProvider {
    @Serializable
    private data class ConfigResponse(
        val jwt: String,
        @SerialName("ice_servers") val iceServers: List<IceServer>? = null,
        val url: String? = null,
    )

    override suspend fun credentials(): RealtimeCredentials {
        val base = backendUrl.trim().trimEnd('/')
        if (base.isEmpty() || runCatching { URL(base) }.isFailure) throw AuthException.InvalidBackendUrl()

        val text = runCatching { Http.request("$base/api/config") }
            .getOrElse { throw AuthException.BackendRequestFailed(it.message ?: "request failed") }
        val config = runCatching { EventJson.decodeFromString<ConfigResponse>(text) }
            .getOrElse { throw AuthException.BackendRequestFailed("invalid config payload") }

        val callsUrl = config.url?.takeIf { it.isNotEmpty() }?.also {
            if (runCatching { URL(it) }.isFailure) {
                throw AuthException.BackendRequestFailed("backend returned an invalid calls url: $it")
            }
        }
        return RealtimeCredentials(
            authorizationHeader = "Bearer ${config.jwt}",
            callsUrl = callsUrl,
            preFetchedIceServers = config.iceServers,
        )
    }
}
